test/units/module_utils/xenserver/conftest.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright: (c) 2019, Bojan Vitnik <bvitnik@mainstream.rs>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+import sys
+import importlib
+import os
+import json
+import pytest
+
+sys.modules['XenAPI'] = importlib.import_module('units.module_utils.xenserver.FakeXenAPI')
```

Comment:
This needs to be in a fixture so it doesn't affect all unit tests. To avoid the need to reference the fixture on each test you may want to use `@pytest.fixture(autouse=True)` on the fixture. Perhaps something like this:

```suggestion
@pytest.fixture(autouse=True)
def fake_xenapi():
    xen_api = importlib.import_module('units.module_utils.xenserver.FakeXenAPI')
    sys.modules['XenAPI'] = xen_api
    return xen_api
```

Reply:
I've experimented with this a little bit and I can make it work but it comes with some ugliness. In most of the test functions I have to do:

`XenAPI = fake_xenapi`

so that `XenAPI` symbol is defined and available locally inside a test function (necessary for mocker.patch()). That's the less ugly part. The more ugly part is that I have to do:

`from ansible.module_utils.xenserver import some_function_to_be_tested`

inside most of the test functions. This import will not work if I put it at the top of the file because at that point `fake_xenapi` fixture is still not called and Python module cache is not populated with FakeXenAPI. As a result of that, `import XenAPI` statement in `ansible.module_utils.xenserver` module will fail and XenAPI symbol will not be available inside `ansible.module_utils.xenserver`.

Puting `from ... import` inside test functions leads to some ugly code repetition. While I'm OK with that, I need to confirm if that is what you had in mind?

Reply:
There is one more thing which is maybe the ugliest. For example, if I have a test function where I forgot to reference `fake_xenapi` fixture properly, `import XenAPI` statement inside `ansible.module_utils.xenserver` will fail. Now imagine if that test function is called first by pytest. As it seems that Python also caches the import failure, the import will also fail for all the other tests. On the other hand, if proper test function is called first by pytest, then the import will work for all the other tests. This makes the outcome of the tests nondeterministic and it can randomly fail depending on the order in which the tests are being run.

Reply:
You shouldn't need to assign `XenAPI = fake_xenapi`. What mocker.patch calls are you trying to make that require that?

To have access to `ansible.module_utils.xenserver` in tests without importing it in each test function, you could wrap it in a fixture.

Reply:
If you're using autouse on your import then you won't need to reference the fixture for it to inject the XenAPI import. Forgetting it will only mean you won't have a local reference to it.

Reply:
For example, in most of the test functions I have:

`>       mocked_xenapi = mocker.patch.object(XenAPI.Session, 'xenapi', create=True)`
`E       NameError: global name 'XenAPI' is not defined`

if I don't do `XenAPI = fake_xenapi`.

Regarding autouse, that could do the trick but I'm afraid that there would be other ways to brake it (not that anything is coming to my mind at the moment). I'll need to experiment more.

Reply:
In that particular example you should be able to use this instead:

```python
mocked_xenapi = mocker.patch.object(fake_xenapi.Session, 'xenapi', create=True)
```

Reply:
You could even rename the fixture to something other than `fake_xenapi` if desired.

Reply:
After some thought, I'd rather do `XenAPI = fake_xenapi` just for the sake of clarity then use `fake_xenapi.Session`. On the other hand, `fake_xenapi` symbol is not available inside a test function if it is not specified as function parameter. `autouse=True` is not enough on it's own. My understanding is that `autouse=True` is OK for fixtures that do not return a value or you don't care about return value. But anyway that doesn't bother me too much.

What bothers me now is that if I wrap both `FakeXenAPI` and `ansible.module_utils.xenserver` in their respective fixtures, I'm not sure that there is a guarantee that a fixture for `FakeXenAPI` will be called first (as it has to be) followed by fixture for `ansible.module_utils.xenserver`. So again we have that problem with ordering of tests and imports.

At this point I fail to understand what is the real benefit of wrapping `FakeXenAPI` inside a fixture. What real problem is solved by doing that? Can you elaborate more on this topic? You said that "This needs to be in a fixture so it doesn't affect all unit tests". If by "all unit tests" you meant all tests in `test/units/module_utils/xenserver` than affecting all tests should be desirable. If you instead meant all unit tests of ansible, then a fixture will not isolate them. My understanding is that imports inside functions (fixture in this case) are still global during a single run of Python program (pytest in this case). Once FakeXenAPI is imported, it will be used from Python module cache for the rest of the pytest session. I'm not sure there is a proper way to remove the module from the cache.

Reply:
Use whatever fixture names make sense. Personally I'd probably name the fixture `XenAPI` since that's the import it is replacing. Then there's no need to assign `XenAPI = fake_xenapi`.

Setting `autouse=True` only loads the fixture for the test automatically. If you want access to that fixture as a local, it still needs to be a function parameter. The benefit to autouse is that it guarantees the import will be mocked, even if the test doesn't require a reference to the fixture. However, there's nothing wrong with specifying an autouse fixture as a function parameter. Doing so simply makes the auto-instantiated fixture available as a local within the test function.

To guarantee correct fixture order, declare that the `fake_xenapi` fixture is used by the fixture providing `ansible.module_utils.xenserver`.

Modifying `sys.modules`, or making other changes to global state, outside of a test fixture or function will apply those change to all unit tests (everything under `test/units/`) when pytest collects the tests before they are executed. Since we run unit tests with the pytest `--boxed` option, each test is run in a separate Python process after the initial collection is performed. This means that any global state changes made in-process by a test function are discarded at the end of that test function, but all changes made during collection are shared.

Imports handled during collection will result in changes to `sys.modules` that will be shared across all tests. This is fine for regular imports. However, the mocked import for `XenAPI` is specific to the tests in `test/units/module_utils/xenserver/`, so it should be restricted to those tests. If another test, such as one for a xenserver module were to also mock `XenAPI` they would conflict.

Reply:
Aaaaaah. Thanks for clarifying the process of test collection, test run and the `--boxed` option. Of course, it makes perfect sense.

As for the order of `fake_xenapi` and `ansible.module_utils.xenserver` fixtures, you are right. Making fixture `ansible.module_utils.xenserver` depend on `fake_xenapi` will force the proper order. In fact, `ansible.module_utils.xenserver` fixture must depend on `fake_xenapi` to even work, and I already implemented it that way, but didn't realize it comes with a bonus of forcing the proper order.

I will make proper changes to the testing code according to everything discussed here. Thanks for  all the comments. I hope I didn't take too much of your time.